libpromises/evalfunction.c

```diff
@@ -6942,48 +6942,45 @@ static FnCallResult FnCallStringSplit(ARG_UNUSED EvalContext *ctx,
 
 static FnCallResult FnCallFileSexist(EvalContext *ctx, ARG_UNUSED const Policy *policy, ARG_UNUSED const FnCall *fp, const Rlist *finalargs)
 {
-    char naked[CF_MAXVARSIZE];
-    char *listvar = RlistScalarValue(finalargs);
-
-    if (IsVarList(listvar))
-    {
-        GetNaked(naked, listvar);
-    }
-    else
-    {
-        Log(LOG_LEVEL_VERBOSE, "Function filesexist was promised a list called '%s' but this was not found", listvar);
-        return FnFailure();
-    }
+    bool allocated = false;
+    JsonElement *json = VarNameOrInlineToJson(ctx, fp, finalargs, false, &allocated);
+    bool ret;
 
-    VarRef *ref = VarRefParse(naked);
-    DataType input_list_type;
-    const Rlist *input_list = EvalContextVariableGet(ctx, ref, &input_list_type);
-    VarRefDestroy(ref);
-    if (input_list_type == CF_DATA_TYPE_NONE)
+    // we failed to produce a valid JsonElement, so give up
+    if (json == NULL)
     {
         Log(LOG_LEVEL_VERBOSE,
-            "Function filesexist was promised a list called"
-            " '%s' but this was not found", listvar);
+            "Cannot produce valid JSON from the argument '%s' of the function '%s'",
+            fp->name, RlistScalarValueSafe(finalargs));
         return FnFailure();
     }
-    else if (DataTypeToRvalType(input_list_type) != RVAL_TYPE_LIST)
+    else if (JsonGetElementType(json) != JSON_ELEMENT_TYPE_CONTAINER)
     {
-        Log(LOG_LEVEL_VERBOSE,
-            "Function filesexist was promised a list called"
-            " '%s' but this variable is not a list", listvar);
+        Log(LOG_LEVEL_VERBOSE, "Function '%s', argument '%s' was not a data container or list",
+            fp->name, RlistScalarValueSafe(finalargs));
+        JsonDestroyMaybe(json, allocated);
         return FnFailure();
     }
 
-    for (const Rlist *rp = input_list; rp != NULL; rp = rp->next)
+    JsonIterator iter = JsonIteratorInit(json);
+    const JsonElement *el = JsonIteratorNextValueByType(&iter, JSON_ELEMENT_TYPE_PRIMITIVE, true);
+
+    /* no elements mean 'false' should be returned, otherwise let's see if the files exist */
+    ret = el != NULL;
```

Comment:
Declare `bool ret` on this line. We use C99 where it makes sense. EDIT: no need to check ret here it seems. Just initiatlise it to true, and name it accordingly (like `file_found`).

Reply:
Done. We need to check if ``el != NULL`` because ``filesexist({})`` needs to be ``false``.